icepap/tcp.py

```diff
@@ -0,0 +1,267 @@
+    errno.ECONNABORTED: ConnectionAbortedError,
+    errno.EPIPE: BrokenPipeError,
+    errno.EBADF: OSError,
+}
+
+
+def to_error(err):
+    if err:
+        return ERR_MAP.get(err, ConnectionError)(err, os.strerror(err))
+
+
+def create_connection(host, port):
+    sock = socket.socket()
+    sock.setblocking(False)
+    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+    res = sock.connect_ex((host, port))
+    if res not in {0, errno.EINPROGRESS}:
+        raise to_error(res)
+    return sock
+
+
+def wait_open(sock, timeout=None):
+    if timeout is None:
+        _, w, _ = select.select((), (sock,), ())
+    elif timeout >= 0:
+        _, w, _ = select.select((), (sock,), (), timeout)
+    else:
+        raise Timeout("timeout trying to connect")
+    if w:
+        err = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
+        if err:
+            raise to_error(err)
+    else:
+        raise Timeout("timeout trying to connect")
+
+
+def stream(sock, buffer_size=8096, timeout=None):
+    readers = sock,
+    while True:
+        start = time.monotonic()
+        r, _, _ = select.select(readers, (), (), timeout)
+        end = time.monotonic()
+        if timeout is not None:
+            timeout -= start - end
+        if (timeout is not None and timeout <= 0) or not r:
+            raise Timeout("read timeout")
+        data = sock.recv(buffer_size)
+        if not data:
+            break
+        yield data
+
+
+def check_open(f):
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        state = self._state
+        if state is CLOSED:
+            raise to_error(errno.EBADF)
+        elif state is OPENING:
+            self.wait_open()
+        return f(self, *args, **kwargs)
+    return wrapper
+
+
+def close_on_error(f):
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        try:
+            return f(self, *args, **kwargs)
+        except BaseException:
+            self.close()
+            raise
+    return wrapper
+
+
+class RawTCP:
+
+    def __init__(self, host, port, eol=b"\n", timeout=None):
+        self.eol = eol
+        self.timeout = timeout
+        self._buffer = b""
+        self._sock = None
+        # create a non blocking socket
+        self._state = OPENING
+        self._connection_time = time.monotonic()
+        self._sock = create_connection(host, port)
+
+    def __del__(self):
+        self.close()
+
+    def wait_open(self):
+        if self._state is OPEN:
+            return
+        elif self._state is CLOSED:
+            raise OSError("would block forever")
+        elif self._state is OPENING:
+            timeout = self.timeout
+            if timeout is not None:
+                timeout -= time.monotonic() - self._connection_time
+                timeout = max(0, timeout)
+            wait_open(self._sock, timeout=timeout)
+        self._state = OPEN
+
+    @close_on_error
+    def _write(self, data):
+        self._sock.sendall(data)
+
```

Comment:
Instead of `sendall()` we should make a loop using the `send()` system call and checking for the timeout as it is done in the `_read()` method right below.

Reply:
Great! 

The change is not necessary if python implements it.